### ISSUE TITLE
[BUG] "Fatal: Dovecot is already running with PID ..."

### ISSUE THREAD

**Issue by coldfix** [docker-mailserver/docker-mailserver#2578]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

dovecot

### What happened and when does this occur?

```Markdown
I got hundreds of Megabytes of log-messages in `mail.log` and consorts containing
the same message (see below). I've had weeks with 1.6 or even 2.3 Million messages.
That's 2-3 messages every second for the whole week.

In some weeks there are only few or even none of these messages for a whole week.
I'm not sure of the exact circumstances. I reckon it might have something to do with
supervisor not recognizing that dovecot is already running.

Interestingly, I don't seem to have any problems with the service other than tons of log
messages. I can send and receive email without problems using a mail client.
```

### What did you expect to happen?

```Markdown
If the system is running as intended, there should be no error messages.
```

### How do we replicate the issue?

```Markdown
Not sure. I basically just `docker-compose up` the service and I can't say when exactly it is happening.
```

### DMS version

v10.5.0

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### What container orchestration tool are you using?

Docker Compose

### docker-compose.yml

```yml
mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    restart: always
    hostname: XXX
    domainname: YYY.de
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
    volumes:
      - ./var/mail/conf:/tmp/docker-mailserver
      - ./var/mail/data:/var/mail
      - ./var/mail/state:/var/mail-state
      - ./var/mail/log:/var/log/mail
      - /etc/localtime:/etc/localtime:ro
      - /etc/letsencrypt:/etc/letsencrypt
    environment:
      # services:
      - ENABLE_CLAMAV=1         # Antivirus
      - ENABLE_FAIL2BAN=1       # Ban IPs after 3 failed login attempts
      - ENABLE_POSTGREY=1
      - ENABLE_SPAMASSASSIN=1
      - ENABLE_QUOTAS=0
      - ENABLE_POP3=0
      # options:
      - POSTGREY_DELAY=50
      - SPAMASSASSIN_SPAM_TO_INBOX=1
      - POSTFIX_MESSAGE_SIZE_LIMIT=134217728    # 128 MiB
      - PFLOGSUMM_TRIGGER=logrotate
      - LOGWATCH_INTERVAL=weekly
      - LOGROTATE_INTERVAL=monthly
      - ONE_DIR=1
      - SSL_TYPE=letsencrypt
    stop_grace_period: 1m
    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
```

### Relevant log output

My logwatch of a recent week reads:

 ```
 --------------------- Dovecot Begin ------------------------ 
 Dovecot Fatal Errors:
    master: Fatal: Dovecot is already running with PID 3881495 (read from /run/dovecot/master.pid): 1677776 Time(s)
 
 Dovecot Deliveries: 122
 Dovecot IMAP and POP3 Successful Logins: 310
 
 Dovecot disconnects: 916 Total
 
 Unknown users blocked: 40 Total
 
 ---------------------- Dovecot End ------------------------- 
```

Sometimes this starts happening directly after a message like this:

```
dovecot: master: Warning: Killed with signal 15 (by pid=3772 uid=0 code=kill)
```

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am uncomfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by georglauterbach** [comment]
When I redeploy my mail server in K8s, I see similar messages for a while, but not exactly the same and not as many :D I will flag this properly, maybe someone has an idea.

**Comment by polarathene** [comment]
You've been running fine for some time without this issue?

Can you confirm your version, and if you updated around the time you noticed this behaviour? Your docker-compose should ideally pin a more specific release [rather than `:latest`](https://vsupalov.com/docker-latest-tag/). Probably won't make a difference, but also try running without `domainname`:

```yaml
mailserver:
    image: docker.io/mailserver/docker-mailserver:11.0
    restart: always
    hostname: XXX.YYY.de
```

How often is [a change detected](https://github.com/docker-mailserver/docker-mailserver/blob/777d4a8cb7da459e36a9c6844273464871350f0c/target/scripts/check-for-changes.sh#L63) in logs? There should be a log at `info` log level (default). Just curious if that behaviour is erratic for some reason, or working as expected, which shouldn't be too often. When a change is detected it will restart Dovecot via `supervisorctl`:

https://github.com/docker-mailserver/docker-mailserver/blob/777d4a8cb7da459e36a9c6844273464871350f0c/target/scripts/check-for-changes.sh#L130

https://github.com/docker-mailserver/docker-mailserver/blob/777d4a8cb7da459e36a9c6844273464871350f0c/target/supervisor/conf.d/supervisor-app.conf#L63-L70

I don't know if there would be any benefit in applying the wrapper approach used for Postfix ([original PR](https://github.com/docker-mailserver/docker-mailserver/pull/676/files#diff-d4fc04cb8f41fc09a2d8a6f5bb9661c69bcc8b9e6be86132ac2674aac825f219)) to Dovecot, but maybe it's useful to help troubleshoot the issue on your end?

---

We would really need to know where the repeated attempts to start Dovecot is coming from. Be sure to properly stop the containers with `docker-compose down` and then `docker-compose up -d` to ensure you're not resuming an old container (which retains state from last time it was up).

Presumably it's something specific to your configuration shared. When we have an easier way to reproduce the problem ourselves, then we can assist at tracking it down :sweat_smile: 

I used your `docker-compose.yml` with a basic setup and didn't see any issues piling up in logs.

**Comment by coldfix** [comment]
Hi,

thanks for your reply.

> Can you confirm your version, and if you updated around the time you noticed this behaviour?

I can confirm I got it on v10.5. I migrated to 11.0 now, and haven't seen the message right now, but I'll let you know if it turns up again.

> Your docker-compose should ideally pin a more specific release [rather than :latest](https://vsupalov.com/docker-latest-tag/).

You're right, that would also make it easier for me to trace when I upgraded - which I'm currently not exactly sure of. Is there an easy way to see this from the log files?

> How often is [a change detected](https://github.com/docker-mailserver/docker-mailserver/blob/777d4a8cb7da459e36a9c6844273464871350f0c/target/scripts/check-for-changes.sh#L63) in logs? There should be a log at info log level (default).

Haven't seen this message at all.

> We would really need to know where the repeated attempts to start Dovecot is coming from. Be sure to properly stop the containers with `docker-compose down` and then `docker-compose up -d` to ensure you're not resuming an old container (which retains state from last time it was up).

Thanks. That might even have been it. So far, I basically did `docker-compose pull && docker-compose up -d `, although I did use `-force-recreate` which should turn out similar, or does it not?

One problem is that I'm right now unable to trace back to the first occurence because my logs only go so far.

> Presumably it's something specific to your configuration shared. When we have an easier way to reproduce the problem ourselves, then we can assist at tracking it down :sweat_smile:

I have only very minimal config for setting up some hard-wired accounts and catchall addresses.


I'll close this issue for now and reopen it if I encounter the problem again, hopefully with better logging capabilities then.

Best, and thanks
Thomas

**Comment by polarathene** [comment]
> Is there an easy way to see this from the log files?

The very first log output when starting a container should be the version of `docker-mailserver`.

> although I did use `--force-recreate` which should turn out similar, or does it not?

That also clears state in the container that isn't persisted to volumes/mounts :+1: 

> I'll close this issue for now and reopen it if I encounter the problem again, hopefully with better logging capabilities then.

Great :)

If it does happen again, we'll need some way to reproduce it on our end to better track down the cause unfortunately. Especially since it doesn't seem to have anything to do with the change detector.

One possible alternative cause for you to consider is memory usage on the system. If available memory at some point gets low, OOM trigger may be killing Dovecot (and perhaps other processes) to try free memory up again... which could be the case if it's doing that back and forth repeatedly.

Maybe also keep an eye on disk space, I've had a server break in the past because of docker log spam with no limits filling up the disk.

**Comment by kolserdav** [comment]
I have the same error with version 14.0.0

**Comment by polarathene** [comment]
@kolserdav ensure you run a fresh container instance `docker compose up --force-recreate`.

If that doesn't work, commenting out your mail-state volume should. You could then find the PID file and delete it, or if there isn't anything important to you in the [`mail-state` volume](https://docker-mailserver.github.io/docker-mailserver/v14.0/config/advanced/optional-config/#volumes-state), just empty the whole volume.

**Comment by Suessiphus** [comment]
Hey there,
in my case the messages occured quite reproducibly after restarting the virtual machine docker runs in (in my case additional in a snap environment, as I used a prepared Ubuntu server with docker in Proxmox) and is related to the short default stop time of 30 s the container had before it was shut down. Hence shutdown was incomplete.
If this happens, stopping and restarting dovecot service inside docker helped in my case to get rid of the messages for the moment.

For the longer term, I changed the timeout to 120 s, which in my case seems enough to complete proper shutdown.
`sudo systemctl edit snap.docker.dockerd.service`
and add 
`[Service]`
`TimeoutStopSec=120`
as proposed in [https://forum.snapcraft.io/t/stop-timeout-not-override-able/14216](url)  and linked descriptions.
